ansible/dualtor/y_cable_simulator_client.j2

```diff
@@ -0,0 +1,157 @@
+from urllib import request, error
+import json
+from sonic_py_common import logger
+
+DUTS_MAP = {
+    "{{ dual_tor_facts['positions']['upper'] }}": 0,
+    "{{ dual_tor_facts['positions']['lower'] }}": 1
+}
+
+VM_SET = "{{ testbed_facts['group-name'] }}"
+
+DUT_NAME = "{{ dut_name }}"
+
+BASE_URL = "http://{{ mux_simulator_server }}:{{ mux_simulator_port }}/"
+
+SYSLOG_IDENTIFIER = "y_cable_sim"
+helper_logger = logger.Logger(SYSLOG_IDENTIFIER)
+
+TOR_A = "tor_a"
+TOR_B = "tor_b"
+
+def _url(port):
+    """
+    Helper function to build an url for given port and target
+    
+    Args:
+        port: physical port on switch, an integer
+    Returns:
+        The url for post/get. 
+    """
+    return BASE_URL + "/mux/{}/{}".format(VM_SET, port - 1)
+
+def _post(port, data):
+    """
+    Helper function for posting data to y_cable server.
+
+    Args:
+        port: physical port on switch, an integer
+        target: TOR_A / TOR_B
+        data: data to post
+    Returns:
+        True if succeed. False otherwise
+    """
+    data = json.dumps(data).encode(encoding='utf-8')
+    header = {'Accept': 'application/json', 'Content-Type': 'application/json'}
+    req = request.Request(url=_url(port), data=data, headers=header)
+    try:
+        res = request.urlopen(req)
+    except error.HTTPError as e:
+        try:
+            err_msg = json.loads(e.read().decode())['err_msg']
+        except Exception:
+            helper_logger.log_warning("post request returns err. status_code = {}".format(e.code))
+        else:
+            helper_logger.log_warning("post request returns err. status_code = {} err_msg = {}".format(e.code, err_msg))
+        return False
+    except error.URLError as e:
+        helper_logger.log_warning("post request returns err. err_msg = {}".format(str(e)))
+        return False
+    return True
+
+def _get(port):
+    """
+    Helper function for polling status from y_cable server.
+
+    Args:
+        port: physical port on switch, an integer
+    Returns:
+        A dict decoded from server's response. Returns None is error is detected.
+    """
+    req = request.Request(url=_url(port))
+    try:
+        res = request.urlopen(req)
+    except error.HTTPError as e:
+        try:
+            err_msg = json.loads(e.read().decode())['err_msg']
+        except Exception:
+            helper_logger.log_warning("get request returns err. status_code = {}".format(e.code))
+        else:
+            helper_logger.log_warning("get request returns err. status_code = {} err_msg = {}".format(e.code, err_msg))
+        return None
+    except error.URLError as e:
+        helper_logger.log_warning("get request returns err. err_msg = {}".format(str(e)))
+        return None
+    data = res.read()
+    return json.loads(data)
```

Comment:
Move the data reading and parsing to right after `res = request.urlopen(req)`? It's because exception could be raised while parsing data as json. Then add another exception handler to catch that.